src/services/endpoints/pattern.service.js

```diff
@@ -0,0 +1,134 @@
+				})
+				.catch((err) => {
+					this._debug('promise rejected', err);
+					util.rejectOnFailure(err.toString(), callback, reject);
+				});
+		});
+	}
+
+	/**
+	 * readAll describes a list of endpoint.
+	 * @param params
+	 * @returns {Promise}
+	 */
+	list(params = {}, callback) {
+		return new Promise((resolve, reject) => {
+			this._debug('initiating a new readAll request');
+			const url = `${this._basePath}?accountId=${params.accountId}&environmentId=${params.environmentId}`;
+			const req = this._client._newRequest('GET', url);
+			this._debug('making list request');
+			this._client._requireOK(this._client._doRequest(req))
+				.then((res) => {
+					this._debug('promise resolved');
+					util.resolveOnSuccess(res.response.items, callback, resolve);
+				})
+				.catch((err) => {
+					this._debug('promise rejected', err);
+					util.rejectOnFailure(err.toString(), callback, reject);
+				});
+		});
+	}
+
+	/**
+	 * create a endpoint.
+	 * @param params
+	 * @returns {Promise}
+	 */
+	create(params = {}, callback) {
+		return new Promise((resolve, reject) => {
+			this._debug('initiating a new create job request, params=', params);
+			this._debug('preparing body');
+			const body = {pattern: Object.assign({}, params)};
+			this._debug('body=', body);
+			const req = this._client._newRequest('POST', `${this._basePath}?accountId=${params.accountId}`, body);
+			this._debug('making create job request');
+			this._client._requireOK(this._client._doRequest(req))
+				.then((res) => {
+					this._debug('promise resolved');
+					util.resolveOnSuccess(res.response.items[0], callback, resolve);
+				})
+				.catch((err) => {
+					this._debug('promise rejected', err);
+					util.rejectOnFailure(err.toString(), callback, reject);
+				});
+		});
+	}
+
+	/**
+	 * update an existing endpoint.
+	 * @param params
+	 * @returns {Promise}
+	 */
+	update(params = {}, callback) {
+		return new Promise((resolve, reject) => {
+			this._debug('initiating a new update request, id=', params.id);
+			this._debug('preparing body');
+			const body = {pattern: Object.assign({}, params)};
+			this._debug('body=', body);
+			delete body.pattern.accountId;
+			delete body.pattern.environmentId;
+			delete body.pattern.id;
+			const req = this._client._newRequest('PUT', `${this._basePath}/${params.id}?accountId=${params.accountId}`, body);
+			this._debug('making update request');
+			this._client._requireOK(this._client._doRequest(req))
+				.then((res) => {
+					this._debug('promise resolved');
+					util.resolveOnSuccess(res.response.items, callback, resolve);
+				})
+				.catch((err) => {
+					this._debug('promise rejected', err);
+					util.rejectOnFailure(err.toString(), callback, reject);
+				});
+		});
+	}
+
+	/**
+	 * delete an existing endpoint.
+	 * @param params
+	 * @returns {Promise}
+	 */
+	delete(params = {}, callback) {
+		return new Promise((resolve, reject) => {
+			if (!util.hasValidResourceId(params.id, callback, reject)) return;
```

Comment:
`util.hasValidResourceId` changed to `util.isValid`, here's an [example](https://github.com/spotinst/spotinst-sdk-nodejs/blob/master/src/services/aws_group_roll.js#L90).